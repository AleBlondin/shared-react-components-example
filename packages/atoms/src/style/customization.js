// @flow

import * as React from 'react';
import { withTheme } from 'styled-components';
import type { Theme } from 'styled-components';

/*
 * This function is used within a component. Its purpose is to overwrite the
 * theme for children widgets.
 */
export const themeMerger = (
  currentComponent: string,
  subThemeKey: string,
  themedComponent: string
) => (parentTheme: Object) => {
  if (
    parentTheme &&
    parentTheme[currentComponent] &&
    parentTheme[currentComponent][subThemeKey]
  ) {
    return {
      ...parentTheme,
      [themedComponent]: {
        ...parentTheme[themedComponent],
        ...parentTheme[currentComponent][subThemeKey],
      },
    };
  }
  return parentTheme || { [themedComponent]: {} };
};

/*
 * This mechanic with several components is necessary due to the fact
 * that only styled-components and components wrapped in 'withTheme' actually
 * receive the theme..
 */
const ComponentThemeBuilder = props => {
  const { render, name, theme, defaultTheme } = props;
  const specificTheme = (theme && theme[name]) || {};
  const finalTheme = {
    ...defaultTheme,
    ...specificTheme,
  };
  return render(finalTheme);
};

const ComponentThemeProvider = withTheme(ComponentThemeBuilder);

ComponentThemeBuilder.defaultProps = {
  theme: {},
};

type ConfigType = {
  defaultTheme: Theme,
  name: string,
};

export const withDefaultTheme = (
  WrappedComponent: React.Node,
  config: ConfigType
) => (props: any) => (
  <ComponentThemeProvider
    defaultTheme={config.defaultTheme}
    name={config.name}
    render={theme => <WrappedComponent {...props} theme={theme} />}
  />
);
