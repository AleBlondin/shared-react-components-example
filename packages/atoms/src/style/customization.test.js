import React from 'react';
import renderer from 'react-test-renderer';
import styled, { ThemeProvider } from 'styled-components';

import { withDefaultTheme, themeMerger } from './customization';

const emptyTheme = {
  MySubComponent: {},
};

describe('themeMerger', () => {
  it('it should return an empty theme for MySubComponent if main theme is null', () => {
    const result = themeMerger('MyComponent', 'subcomponent', 'MySubComponent')(
      null
    );
    expect(result).toEqual(emptyTheme);
  });

  it('it should return an empty theme for MySubComponent if main theme is undefined', () => {
    const result = themeMerger(
      'MyComponent',
      'subcomponent',
      'MySubComponent'
    )();
    expect(result).toEqual(emptyTheme);
  });

  it('it should return main theme if there is no specific theme for the widget', () => {
    const inputTheme = {
      MySubComponent: {
        color: 'orange',
      },
    };
    const result = themeMerger('MyComponent', 'subcomponent', 'MySubComponent')(
      inputTheme
    );
    expect(result).toEqual(inputTheme);
  });

  it('it should return main theme if the widget theme does not define any specific theme for the subcomponent', () => {
    const inputTheme = {
      MyComponent: {
        color: 'black',
      },
      MySubComponent: {
        color: 'orange',
      },
    };
    const result = themeMerger('MyComponent', 'subcomponent', 'MySubComponent')(
      inputTheme
    );
    expect(result).toEqual(inputTheme);
  });

  it('it should copy the specific theme for the subcomponent into the main theme', () => {
    const inputTheme = {
      MyComponent: {
        color: 'black',
        subcomponent: {
          color: 'red',
        },
      },
    };
    const expectedTheme = {
      MyComponent: {
        color: 'black',
        subcomponent: {
          color: 'red',
        },
      },
      MySubComponent: {
        color: 'red',
      },
    };
    const result = themeMerger('MyComponent', 'subcomponent', 'MySubComponent')(
      inputTheme
    );
    expect(result).toEqual(expectedTheme);
  });

  it('it should copy the specific theme for the subcomponent into the main theme and merge with the existing one', () => {
    const inputTheme = {
      MyComponent: {
        color: 'black',
        subcomponent: {
          color: 'red',
        },
      },
      MySubComponent: {
        color: 'orange',
        otherProp: 'value',
      },
    };

    const expectedTheme = {
      MyComponent: {
        color: 'black',
        subcomponent: {
          color: 'red',
        },
      },
      MySubComponent: {
        color: 'red',
        otherProp: 'value',
      },
    };

    const result = themeMerger('MyComponent', 'subcomponent', 'MySubComponent')(
      inputTheme
    );
    expect(result).toEqual(expectedTheme);
  });
});

const theme = {
  MyComponent: {
    color: 'black',
    subcomponent: {
      color: 'red',
    },
  },
  MySubComponent: {
    color: 'orange',
    otherProp: 'value',
  },
};
const defaultTheme = {
  color: 'red',
  backgroundColor: 'black',
};

const StyledComponent = styled.div`
  color: ${props => props.theme.color};
  background-color: ${props => props.theme.backgroundColor};
`;
const WrappedComponent = withDefaultTheme(StyledComponent, {
  defaultTheme,
  name: 'MySubComponent',
});

describe('withDefaultTheme', () => {
  it('should return a component', () => {
    const component = renderer.create(
      <ThemeProvider theme={theme}>
        <WrappedComponent />
      </ThemeProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should return a component with the default theme', () => {
    const component = renderer.create(<WrappedComponent />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
