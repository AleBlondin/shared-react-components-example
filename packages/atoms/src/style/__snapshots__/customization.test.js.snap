// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withDefaultTheme should return a component 1`] = `
.c0 {
  color: orange;
  background-color: black;
}

<div
  className="c0"
/>
`;

exports[`withDefaultTheme should return a component with the default theme 1`] = `
.c0 {
  color: red;
  background-color: black;
}

<div
  className="c0"
/>
`;
